### DISCUSSION TITLE
[discussion] (PHP代码实现的功能能不能单列出来)

### DISCUSSION THREAD

**Discussion by rwcoding** [swow/swow#95] [General]
1.  `lib` 库里PHP代码的功能和扩展实现的功能都在一个命名空间下，搞不清楚是扩展提供的还是PHP库提供的
2.  如果开发者不喜欢PHP库的一些实现方式，它就显得有点多余。只要有足够健壮高效的底层，大家都喜欢造上层轮子
3.  为什么不作为单独的官方库实现，可以composer引入使用，也可以作为框架开发者的参考

**Comment by twose** [comment]
1. 因为不希望开发者去关注到底是底层实现还是PHP库实现
2. 正是因为喜欢用PHP开发而不是C开发扩展才会这么做
3. 只提供基础的功能，框架应该在此之上而不是重新实现
4. PHP内核未来也会使用PHP代码实现部分C的内容， 当然，区别是 它不是通过composer提供

**Comment by AuroraYolo** [comment]
@twose 其实不太明白的是，为什么重复的造轮子也只是实现同样的基础功能。难道比的是那个轮子造的比较好吗？实现的更优雅吗？

**Comment by twose** [comment]
> @twose 其实不太明白的是，为什么重复的造轮子也只是实现同样的基础功能。难道比的是那个轮子造的比较好吗？实现的更优雅吗？

这里的轮子是指什么?

**Comment by AuroraYolo** [comment]
比如框架的组件里的http,client,或者response。太多相似的代码实现相同的功能.

**Comment by twose** [comment]
@codingheping 框架里的HTTP client也可以作为其它库的驱动比如guzzle, 这也是遵循PSR的意义所在, 他们并不是互斥的
当然我也推荐直接使用curl hook

**Reply by he426100** [reply]
为什么http client会比curl快的？ 
github codespace环境下请求 `https://ipinfo.io/json`，curl 0.46，client 0.22。
问题是client不支持Https？有些网站需要connect 80，有些需要443，但是连443会报400错误，有办法解决么？

**Reply by he426100** [reply]
试了下swoole的http client，跟swow的速度一样，都比curl快很多

**Reply by twose** [reply]
> 试了下swoole的http client，跟swow的速度一样，都比curl快很多

支持，需要在`connect`后调用`enableCrypto`来开启`SSL`，因为有`cURL`支持，暂时还没有做傻瓜式的`Client`。

或者你可以试试`file_get_contents()`，也很快。
